### PR TITLE
fix(upgrade): fix several small issues

### LINKED ISSUE #18913: [ngUpgrade] onInit fires too early on downgraded components
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

The following component is downgraded and used in an AngularJS application:

```typescript
export class Ng2ComponentBindings implements OnInit, OnChanges {
  @Input() foo: {a: number};

  ngOnInit() {
    console.log('ngInit:', this.foo);
  }

  ngOnChanges() {
    console.log('ngOnChange:', this.foo);
  }
}

ng1module.directive(
  'ng2ComponentBindings',  // lowerCamel when registering.
  downgradeComponent(
      {component: Ng2ComponentBindings, propagateDigest: true}));
```

it is used like this ``` <ng2-component-bindings [foo]="$ctrl.foo"></ng2-component-bindings>```

The order of events is:
- ngOnInit fires
- foo is set
- ngOnChanges fires

## Expected behavior

The order of events should be:
- foo is set
- ngOnChanges fires
- ngOnInit fires

Because that is how a regular Angular component would behave.

### PR DESCRIPTION
**Pull request by gkalpak**
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~

## PR Type
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
- [ngUpgradeFull] AngularJS is not bootstraped inside the Angular zone in tests (as happens in real applications).
- [ngUpgradeFull] Angular change detection is not wired up correctly when `propagateDigest` is false.
- [ngUpgradeLite] Downgraded components unnecessarily call `appRef.detach()` on destroy.
- [ngUpgradeLite] Downgraded components are not destroyed inside the Angular zone.
- [ngUpgradeLite] `ngOnInit()` is called before `ngOnChanges()` (on components with inputs).

## What is the new behavior?
The above issues are fixed.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
Fixes #18913.

### CODE REVIEW

File: packages/upgrade/src/common/downgrade_component_adapter.ts [hunk 1]
```diff
@@ -152,9 +149,16 @@ export class DowngradeComponentAdapter {
       this.componentScope.$watch(this.wrapCallback(detectChanges));
     }
 
-    // Attach the view so that it will be dirty-checked.
-    if (needsNgZone) {
-      this.appRef.attachView(this.componentRef.hostView);
+    // If necessary, attach the view so that it will be dirty-checked.
+    // (Allow time for the initial input values to be set and `ngOnChanges()` to be called.)
+    if (needsNgZone || !propagateDigest) {
```
Comment (gkalpak) [inline packages/upgrade/src/common/downgrade_component_adapter.ts]: As discusssed [here][1].

[1]: https://github.com/angular/angular/pull/18209#discussion_r128622579

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts [hunk 2]
```diff
@@ -270,6 +306,210 @@ export function main() {
...
+
+               // Destroy the component.
+               rootScope.$apply('hideNg2 = true');
+               expect(logs).toEqual([
+                 'OnDestroy(qux)',
+               ]);
+               logs.length = 0;
+             });
+           });
+         }));
```
Comment (gkalpak) [inline packages/upgrade/test/static/integration/downgrade_module_spec.ts]: As shown in this test, due to `ngZone.run()` causing an `onMicrotaskEmpty` emission, there are more change detection runs than necessary in some cases.

With `propagateDigest: true`, there can be up to two extra runs (when inputs change during a `$digest` run from outside the Angular zone).
With `propagateDigest: false`, there can be up to one extra run (when inputs change during a `$digest` run from outside the Angular zone).

Note that this behavior existed before this PR - it is just highlighted more by this test.

##
I am not sure if/how we could fix this (without introducing some asynchronicity that might break apps). Hopefully, in the most performance-critical setup (`downgradeModule()` + `propagateDigest: false`) it only adds one extra run for each time the inputs change (during a `$digest` run outside the Angular zone).

File: packages/upgrade/test/static/test_helpers.ts [hunk 3]
```diff
@@ -18,11 +18,19 @@ export function bootstrap(
...
+    const ngZone = ref.injector.get<NgZone>(NgZone);
     const upgrade = ref.injector.get(UpgradeModule);
     const failHardModule: any = ($provide: angular.IProvideService) => {
       $provide.value('$exceptionHandler', (err: any) => { throw err; });
     };
-    upgrade.bootstrap(element, [failHardModule, ng1Module.name]);
+
+    // The `bootstrap()` helper is used for convenience in tests, so that we don't have to inject
+    // and call `upgrade.bootstrap()` on every Angular module.
+    // In order to closer emulate what happens in real application, ensure AngularJS is bootstraped
```
Comment (petebacondarwin) [inline packages/upgrade/test/static/test_helpers.ts]: bootstrap__p__ed

File: packages/upgrade/test/static/test_helpers.ts [hunk 4]
```diff
@@ -18,11 +18,19 @@ export function bootstrap(
...
+    // inside the Angular zone.
+    //
+    ngZone.run(() => upgrade.bootstrap(element, [failHardModule, ng1Module.name]));
```
Comment (petebacondarwin) [inline packages/upgrade/test/static/test_helpers.ts]: Should we have a test for this, or perhaps should a test be failing without it?
Otherwise what are we benefitting from this change?

File: packages/upgrade/test/static/integration/downgrade_component_spec.ts [hunk 5]
```diff
@@ -274,6 +274,42 @@ export function main() {
          });
        }));
 
+    it('should still run normal Angular change-detection regarless of `propagateDigest`',
```
Comment (petebacondarwin) [inline packages/upgrade/test/static/integration/downgrade_component_spec.ts]: regar_d_less

File: packages/upgrade/src/common/downgrade_component_adapter.ts [hunk 6]
```diff
@@ -150,10 +149,16 @@ export class DowngradeComponentAdapter {
...
     }
 
-    // Attach the view so that it will be dirty-checked.
+    // If necessary, attach the view so that it will be dirty-checked.
+    // (Allow time for the initial input values to be set and `ngOnChanges()` to be called.)
     if (needsNgZone || !propagateDigest) {
-      const appRef = this.parentInjector.get<ApplicationRef>(ApplicationRef);
-      appRef.attachView(this.componentRef.hostView);
+      let unwatch: Function|null = this.componentScope.$watch(() => {
+        unwatch !();
```
Comment (petebacondarwin) [inline packages/upgrade/src/common/downgrade_component_adapter.ts]: so ugly!

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts [hunk 7]
```diff
@@ -306,6 +306,162 @@ export function main() {
...
+             @Input() value = 'foo';
+
+             ngAfterContentChecked() { this.log('AfterContentChecked'); }
+             ngAfterContentInit() { this.log('AfterContentInit'); }
+             ngAfterViewChecked() { this.log('AfterViewChecked'); }
+             ngAfterViewInit() { this.log('AfterViewInit'); }
+             ngDoCheck() { this.log('DoCheck'); }
+             ngOnChanges() { this.log('OnChanges'); }
+             ngOnDestroy() { this.log('OnDestroy'); }
+             ngOnInit() { this.log('OnInit'); }
```
Comment (petebacondarwin) [inline packages/upgrade/test/static/integration/downgrade_module_spec.ts]: Didn't fancy doing this by chaining Jasmine spies together? :-P

**Review by petebacondarwin — APPROVED** [review]
LGTM - just a couple of minor typos

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts (hunk 7 above)
Comment (gkalpak) [reply packages/upgrade/test/static/integration/downgrade_module_spec.ts]: 😛 

There is merit in using the `log` approach, when you have multiple calls of each method in different relative orders :grin:
Choose the right tool for the job, right :)

File: packages/upgrade/test/static/test_helpers.ts (hunk 4 above)
Comment (gkalpak) [reply packages/upgrade/test/static/test_helpers.ts]: IIRC [this][1] would fail without this change.

The main idea was to ensure that tests still work when run inside the zone (similar to what happens in real apps). I don't feet too compelled to test the testing helper :wink:

[1]: https://github.com/angular/angular/pull/19394/commits/732084de6454fa98d6f5ed204b851dd8ec4752fb#diff-a7f6b73480141da01c59b673f1ba5f27R446
